pkg/driver/aci_driver.go

```diff
@@ -0,0 +1,760 @@
+	}
+
+	return nil, fmt.Errorf("ACI_MSI_TYPE environment variable unknown value: %s ", useMSI)
+}
+
+func (d *ACIDriver) getRoleDefinitionsClient(subscriptionID string) authorization.RoleDefinitionsClient {
+	roleDefinitionsClient := authorization.NewRoleDefinitionsClient(subscriptionID)
+	roleDefinitionsClient.Authorizer = d.authorizer
+	roleDefinitionsClient.AddToUserAgent(userAgent)
+	return roleDefinitionsClient
+}
+
+func (d *ACIDriver) getRoleAssignmentClient(subscriptionID string) authorization.RoleAssignmentsClient {
+	roleAssignmentsClient := authorization.NewRoleAssignmentsClient(subscriptionID)
+	roleAssignmentsClient.Authorizer = d.authorizer
+	roleAssignmentsClient.AddToUserAgent(userAgent)
+	return roleAssignmentsClient
+}
+
+func (d *ACIDriver) getUserAssignedIdentitiesClient(subscriptionID string) msi.UserAssignedIdentitiesClient {
+	userAssignedIdentitiesClient := msi.NewUserAssignedIdentitiesClient(subscriptionID)
+	userAssignedIdentitiesClient.Authorizer = d.authorizer
+	userAssignedIdentitiesClient.AddToUserAgent(userAgent)
+	return userAssignedIdentitiesClient
+}
+
+func (d *ACIDriver) getContainerGroupsClient() containerinstance.ContainerGroupsClient {
+	containerGroupsClient := containerinstance.NewContainerGroupsClient(d.subscriptionID)
+	containerGroupsClient.Authorizer = d.authorizer
+	containerGroupsClient.AddToUserAgent(userAgent)
+	return containerGroupsClient
+}
+
+func (d *ACIDriver) getContainerClient() containerinstance.ContainerClient {
+	containerClient := containerinstance.NewContainerClient(d.subscriptionID)
+	containerClient.Authorizer = d.authorizer
+	containerClient.AddToUserAgent(userAgent)
+	return containerClient
+}
+
+func (d *ACIDriver) getGroupsClient() resources.GroupsClient {
+	groupsClient := resources.NewGroupsClient(d.subscriptionID)
+	groupsClient.Authorizer = d.authorizer
+	groupsClient.AddToUserAgent(userAgent)
+	return groupsClient
+}
+
+func (d *ACIDriver) getProvidersClient() resources.ProvidersClient {
+	providersClient := resources.NewProvidersClient(d.subscriptionID)
+	providersClient.Authorizer = d.authorizer
+	providersClient.AddToUserAgent(userAgent)
+	return providersClient
+}
+
+func (d *ACIDriver) getContainerState(aciRG string, aciName string) (string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	containerGroupsClient := d.getContainerGroupsClient()
+	resp, err := containerGroupsClient.Get(ctx, aciRG, aciName)
+	if err != nil {
+		return "", err
+	}
+
+	return *resp.InstanceView.State, nil
+}
+
+func locationIsAvailable(location string, locations []string) bool {
+
+	for _, l := range locations {
+		l = strings.ToLower(strings.Replace(l, " ", "", -1))
+		if l == location {
+			return true
+		}
+	}
+
+	return false
+}
+func checkForMSIEndpoint() bool {
+	timeout := time.Duration(1 * time.Second)
+	client := http.Client{
+		Timeout: timeout,
+	}
+	_, err := client.Head("http://169.254.169.254/metadata/identity/oauth2/token")
```

Comment:
nit: could we have the IP address as a constant, and also indicate what it represents?